Add H5Dread_chunk2()

fortnern opened the pull request:
Version H5Dread_chunk() according to the standard HDF5 API versioning scheme. H5Dread_chunk2() takes an additional parameter that points to a variable containing the size of the buffer on entry, and is set to the needed size on exit. Deprecated H5Dread_chunk1(). The FORTRAN interface still needs to be updated.

Resolves #5224

fortnern committed:
Add H5Dread_chunk2().  Missing testing and documentation.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Some changes to H5Dread_chunk2() assertions, etc.

fortnern committed:
Add documentation and testing for H5Dread_chunk2()

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Add RELEASE.txt entry for H5Dread_chunk2()

brtnfld commented on release_docs/RELEASE.txt:
```diff
@@ -171,6 +171,31 @@ New Features
+
+      A new parameter, nalloc, has been added to H5Dread_chunk(). This parameter
+      contains a pointer to a variable that holds the size of the buffer buf. If
+      *nalloc is not large enough to hol dhte entire chunk being read, no data
```
"hol dhte"

brtnfld replied:
fixed

brtnfld commented on src/H5Dpublic.h:
```diff
@@ -1250,10 +1250,11 @@ H5_DLL herr_t H5Dwrite_chunk(hid_t dset_id, hid_t dxpl_id, uint32_t filters, con
  *                      dataspace
  * \param[in,out]  filters  Mask for identifying the filters in use
  * \param[out]  buf     Buffer containing data to be read from the chunk
+ * \param[in,out]  nalloc   Size of buf in bytes
```
Is the naming convention `nalloc`, or `buf_size`? I see both being used. What should be used going forward? For Fortran, I think we mostly use buf_size.

fortnern replied:
I used nalloc because that's what the closest existing analogs for this pattern, H5Tencode etc., use. I do think buf_size is a better name though, so I wouldn't be opposed to changing it.

brtnfld replied:
I changed it to `buf_size`.

brtnfld commented on src/H5Dpublic.h:
```diff
@@ -1276,22 +1277,31 @@ H5_DLL herr_t H5Dwrite_chunk(hid_t dset_id, hid_t dxpl_id, uint32_t filters, con
+ *          \p *nalloc is set to the buffer size needed to read the chunk, which
+ *          is the same as the size of the chunk on disk. If the value of
+ *          \p *nalloc passed in was insufficient to read the entire, chunk, no
+ *          data is read. \p buf may be passed as NULL as long as \p *nalloc
```
Why does `nalloc` need to be set to zero when `buf ` is NULL?

fortnern replied:
I suppose we could allow a nonzero (uninitialized *nalloc) in this case for convenience. My reasoning was that by passing *nalloc > 0 the app is asserting there is space in the buffer, and therefore it's inconsistent to pass a NULL buffer.

fortnern replied:
Now that I look at it, H5Tencode allows *nalloc > 0 with buf == NULL, so I'll change this.

brtnfld replied:
Ok, then I will allow this behaviour in the FORTRAN wrapper.

brtnfld replied:
done

brtnfld committed:
fortran wrapper and test for H5Dread_chunk

github-actions[bot] committed:
Committing clang-format changes

brtnfld committed:
Merge branch 'develop' into secure_read_chunk

brtnfld commented:
@fortnern , I finished with the Fortran additions, you might want to bring the PR in sync with develop

brtnfld committed:
removed commented out code

brtnfld commented:
@fortnern, I don't think those failing test are related to my changes. All tests pass on my OpenSUSE

fortnern committed:
Merge branch 'develop' into secure_read_chunk

fortnern committed:
Allow nonzero *buf_size with NULL buf
Fix test failures
Add useful error messages to direct_chunk.c
Other test updates/fixes

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Fix uninitialized variable when deflate is disabled.

github-actions[bot] committed:
Committing clang-format changes

epourmal commented on fortran/src/H5Dff.F90:
```diff
@@ -1780,6 +1786,62 @@ SUBROUTINE h5dfill_f(fill_value, fill_type_id, buf, buf_type_id, space_id, hdfer
+    INTEGER(HSIZE_T)  , INTENT(IN), DIMENSION(:) :: offset
+    INTEGER           , INTENT(INOUT) :: filters
+    TYPE(C_PTR)                       :: buf
+    INTEGER(HID_T)    , INTENT(INOUT) :: buf_size
```
Should the type be INTEGER(SIZE_T)?

brtnfld replied:
good catch, fixed.

brtnfld committed:
correct buf_size type in h5dread_chunk_f

fortnern committed:
Merge branch 'develop' into secure_read_chunk

fortnern commented on hl/src/H5DO.c:
```diff
@@ -60,7 +60,7 @@ herr_t
 {
     /* Call underlying H5D function */
-    if (H5Dread_chunk(dset_id, dxpl_id, offset, filters, buf) < 0)
+    if (H5Dread_chunk1(dset_id, dxpl_id, offset, filters, buf) < 0)
```
Thanks copilot, but this code is within an #ifndef H5_NO_DEPRECATED_SYMBOLS block so there should be no issue. Also calling the macro in the library is a bad idea since the signature is different depending on which version gets called.

brtnfld committed:
update link to https

fortnern committed:
Revert fortran/src/CMakeLists.txt to develop

fortnern committed:
Merge branch 'secure_read_chunk' of github.com:fortnern/hdf5 into secure_read_chunk

brtnfld committed:
switch to netcdf-c action with h5dread_chunk2 changes

brtnfld committed:
switch to netcdf-c action with h5dread_chunk2 changes

brtnfld committed:
spacing

brtnfld committed:
update timeout

brtnfld committed:
added Check URLs regardless of any robots.txt files.

fortnern commented on src/H5version.h:
```diff
@@ -368,6 +368,10 @@
   #define H5Dopen_vers 2
 #endif /* !defined(H5Dopen_vers) */
 
+#if !defined(H5Dread_chunk_vers)
+  #define H5Dread_chunk_vers 1
+#endif /* !defined(H5Dread_chunk_vers) */
```
Copilot doesn't understand our API versioning scheme

brtnfld committed:
increase linker check timeout

brtnfld committed:
switch back to netcdf repository and remove link checking changes
